internal/epoch/epoch_manager.go

```diff
@@ -950,6 +1023,101 @@ func (e *Manager) getCompleteIndexSetForCommittedState(ctx context.Context, cs C
 	return result, nil
 }
 
+type intRange struct {
+	lo, hi int
+}
+
+func (r intRange) length() int {
+	return r.hi - r.lo
+}
+
+func (r intRange) isEmpty() bool {
+	return r.length() <= 0
+}
+
+var errNonContiguousRange = errors.New("non-contiguous range")
+
+// Returns a continuous close-open epoch range for the keys, that is [lo, hi).
+// A range of the form [v,v) means the range is empty.
+// When the range is not continuous an error is returned.
+func getKeyRange[E any](m map[int]E) (intRange, error) {
+	keys := maps.Keys(m)
+
+	slices.Sort(keys)
+
+	keys = slices.Compact(keys)
+
+	l := len(keys)
+	if l == 0 {
+		return intRange{}, nil
+	}
+
+	if lo, hi := keys[0], keys[l-1]; hi-lo > l-1 {
+		return intRange{lo: -1, hi: -1}, errors.Wrapf(errNonContiguousRange, "lo: %d, hi: %d, length: %d", lo, hi, l)
+	}
+
+	return intRange{lo: keys[0], hi: keys[l-1] + 1}, nil
+}
+
```

Comment:
I'm not sure how big `m` could be but we can do slightly better w/o `slices.Sort`. The implementation below is more efficient for large `m`: O(n log n) -> O(n). It also removes dependencies on experimental `golang.org/x/exp/maps` package.

```
func getKeyRange[E any](m map[int]E) (intRange, error) {
    lo, hi, count := 0, 0, 0
    for k := range m {
        if count == 0 { // init lo and hi with the 1st key.
            lo, hi = k, k 
        } else {
            if k < lo {
                lo = k
            }
            if k > hi {
                hi = k
            }
        }
        count++
    }
    
    if count == 0 { // no keys were found
        return intRange{}, nil
    }

    if hi-lo > count-1 { // is the range contiguous?
        return intRange{lo: -1, hi: -1}, errors.Wrapf(errNonContiguousRange, "lo: %d, hi: %d, length: %d", r.lo, r.hi, count)
    }

    return intRange{lo: lo, hi: hi+1}, nil
}
```
I didn't test it :)

Reply:
Thanks, nice suggestion!
I'll put this in a separate PR with tests, and then use it here.

Reply:
In general, the length of these ranges, and thus the size of the input maps, is expected to be in the single digits. If the number of uncompacted epochs grows too large, then there will be other performance issues, and the algorithmic complexity of this function will be irrelevant even in those cases.

The suggested implementation with tests is in #3721 @plar @redgoat650 PTAL. This PR (#3651) now depends on that one. Thanks.